pkg/controller/node/status.go

```diff
@@ -60,24 +89,91 @@ func calculateStatus(cconfig *v1.ControllerConfig, pool *mcfgv1.MachineConfigPoo
+	var degradedMachines, readyMachines, updatedMachines, unavailableMachines, updatingMachines []*corev1.Node
+	// if we represent updating properly here, we will also represent updating properly in the CO
+	// so this solves the cordoning RFE and the upgradeable RFE
+	// updating == updatePrepared, updateExecuted, updatedComplete
+	// updated == nodeResumed, updated
+	// ready == nodeResumed, updated
+	// unavailable == draining, cordoned
+	// degraded == if the condition.Reason == error
+	// this ensures that a MCP only enters Upgradeable==False if the node actually needs to upgrade to the new MC
+	for _, state := range mcs {
+		var ourNode *corev1.Node
+		for _, n := range nodes {
+			if state.Name == n.Name {
+				ourNode = n
+				break
+			}
+		}
+		if ourNode == nil {
+			klog.Errorf("Could not find specified node %s", state.Name)
+		}
+		if len(state.Status.Conditions) == 0 {
+			// not ready yet
+			break
+		}
+		for _, cond := range state.Status.Conditions {
+			if cond.Reason == "Error" {
+				degradedMachines = append(degradedMachines, ourNode)
+				continue
+			}
+			if cond.Status == metav1.ConditionUnknown {
+				switch v1alpha1.StateProgress(cond.Type) {
+				case mcfgalphav1.MachineConfigNodeUpdatePrepared:
+					updatingMachines = append(updatedMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateExecuted:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdatePostActionComplete:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateComplete:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeResumed:
+					updatedMachines = append(updatingMachines, ourNode)
+					readyMachines = append(readyMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateCompatible:
+					updatedMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateDrained:
+					unavailableMachines = append(unavailableMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateCordoned:
+					unavailableMachines = append(unavailableMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdated:
+					updatedMachines = append(updatedMachines, ourNode)
+					readyMachines = append(readyMachines, ourNode)
+				}
+			}
+		}
+	}
+	degradedMachineCount := int32(len(degradedMachines))
 	updatedMachineCount := int32(len(updatedMachines))
-
-	readyMachines := getReadyMachines(pool, nodes)
+	unavailableMachineCount := int32(len(unavailableMachines))
+	updatingMachineCount := int32(len(updatingMachines))
 	readyMachineCount := int32(len(readyMachines))
 
-	unavailableMachines := getUnavailableMachines(nodes, pool)
-	unavailableMachineCount := int32(len(unavailableMachines))
+	// this is # 1 priority, get the upgrade states actually reporting
+	if degradedMachineCount+readyMachineCount+unavailableMachineCount+updatingMachineCount != int32(len(nodes)) {
```

Comment:
Similar to my question above, I guess we can only be in one of 4 states now:
1. degraded
2. ready and updated
3. unavailable
4. updating

Can a node, for example, not be unavailable (kubelet not up) and updating (rebooting)? I feel like we used to be able to

Reply:
Put another way it feels like we're calculating two ways of status (if new way doesn't add up, fall back to old way) and there might be some differences in behaviour here

Reply:
This is the biggest weird spot of the entire PR. The way I wanted to represent this was:

1. updated is only true when the daemon is up and running and goes through a normal `syncNode`
2. updating is true in a. updatepreprared b. updateexecuted c. updatecomplete
3. ready and updated I chose to combine since they occur so synchronously. However, Ready indicates a machine that is out of the whole "updating loop" and is about to run normally again
4. unavailable is for cordoning and draining nodes specifically
5. degraded is if we hit an upgrade error. 

I might not have represented that properly, I can check. But this was my intention. I wanted to maintain the MCP updating,updated,degraded,unavailable information but I wanted to augment how those states were chosen WITH mcn data. Does that make sense?

Reply:
Discussed this with Charlie, as long as this doesn't break existing non-featuregated use case, I am fine with that. We can revisit this when bringing it to GA

Reply:
FWIW: I was having some trouble with this when I was implementing the BuildController changes.